staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -1186,6 +1190,36 @@ func schemaIsNonStructural(schema *apiextensions.JSONSchemaProps) bool {
 	return len(structuralschema.ValidateStructural(nil, ss)) > 0
 }
 
+// requirePrunedDefaults returns true if the old CRD spec had no unpruned defaults.
+func requirePrunedDefaults(oldCRDSpec *apiextensions.CustomResourceDefinitionSpec) bool {
+	if oldCRDSpec.Validation != nil && schemaIsNonStructuralOrHasUnprunedDefaults(oldCRDSpec.Validation.OpenAPIV3Schema) {
+		return false
+	}
+	for _, v := range oldCRDSpec.Versions {
+		if v.Schema != nil && schemaIsNonStructuralOrHasUnprunedDefaults(v.Schema.OpenAPIV3Schema) {
+			return false
+		}
+	}
+	return true
+}
+func schemaIsNonStructuralOrHasUnprunedDefaults(schema *apiextensions.JSONSchemaProps) bool {
```

Comment:
this is a confusing name, since this returns false when the schema is non-structural

Reply:
would something like this be simpler?

```go
// schemaMightHaveUnprunedDefaults returns true if the schema contains unpruned defaults,
// or if the schema contains defaults and and an error occurs attempting to determine if they are pruned.
func schemaMightHaveUnprunedDefaults(schema *apiextensions.JSONSchemaProps) bool {
  if schema == nil || !schemaHasDefaults(schema) {
    return false
  }
  ss, err := structuralschema.NewStructural(schema)
  if err != nil {
    // we have defaults with a non-structural schema, they might be unpruned
    return true
  }
  if len(structuralschema.ValidateStructural(nil, ss)) > 0 {
    // we have defaults with a non-structural schema, they might be unpruned
    return true
  }
  pruned := ss.DeepCopy()
  if err := structuraldefaulting.PruneDefaults(pruned); err != nil {
    // an unexpected error occurred attempting to prune defaults
    return true
  }
  return !reflect.DeepEqual(ss, pruned)
}
```

Reply:
then the usage looks like:

```go
func requirePrunedDefaults(oldCRDSpec *apiextensions.CustomResourceDefinitionSpec) bool {
  if schemaMightHaveUnprunedDefaults(oldCRDSpec.Validation.OpenAPIV3Schema) {
    return false
  }
  for _, v := range oldCRDSpec.Versions {
    if schemaMightHaveUnprunedDefaults(v.Schema.OpenAPIV3Schema) {
      return false
    }
  }
  return true
}
```